src/OrleansRuntime/Async/CancellationSourcesExtension.cs

```diff
@@ -0,0 +1,91 @@
+﻿using System;
+using System.Threading.Tasks;
+using Orleans.CodeGeneration;
+using Orleans.Runtime;
+using Orleans.Runtime.Providers;
+
+namespace Orleans.Async
+{
+    /// <summary>
+    /// Contains list of cancellation token source corresponding to the tokens
+    /// passed to the related grain activation.
+    /// </summary>
+    internal class CancellationSourcesExtension : ICancellationSourcesExtension
+    {
+        private readonly Lazy<TraceLogger> _logger = new Lazy<TraceLogger>(() =>
+            TraceLogger.GetLogger("CancellationSourcesExtension", TraceLogger.LoggerType.Application));
+        private static readonly Interner<long, GrainCancellationToken> _cancellationTokens;
+        private static readonly TimeSpan _cleanupFrequency = TimeSpan.FromMinutes(3);
+        private static readonly int _defaultInternerCollectionSize = 31;
+
+        static CancellationSourcesExtension()
+        {
+            _cancellationTokens = new Interner<long, GrainCancellationToken>(
+                _defaultInternerCollectionSize,
+                _cleanupFrequency);
+        }
+
+        public Task CancelTokenSource(uint grainId, GrainCancellationToken token)
```

Comment:
Passing of the `GrainCancellationToken` instead of its id is neccesary because of possible race condition when cancel request arrives before original one. If to pass only id - it will cause extension not installed exception because cancellation extension is being installed when `GCT` is found in parameters list.